### PR TITLE
UCP/PROTO: Infra to print protocols performance

### PR DESCRIPTION
**Pull request by yosefe**
## Why
Add the infrastructure to print protocol performance visualization as .dot file. The output is demonstrated in #8233

### CODE REVIEW

**Comment by yosefe** [comment]
@dmitrygx @brminich can you pls take a look?

File: src/ucp/proto/proto_debug.c [hunk 1]
```diff
@@ -609,6 +610,265 @@ const char *ucp_proto_perf_node_desc(ucp_proto_perf_node_t *perf_node)
...
+        } else {
+            ucs_string_buffer_appendf(strb, "%.1fGBs", mbps / 1024);
+        }
+        if (data->value.c > UCP_PROTO_PERF_EPSILON) {
+            ucs_string_buffer_appendf(strb, "+");
+        }
+    }
+
+    if (data->value.c > UCP_PROTO_PERF_EPSILON) {
+        ucs_string_buffer_appendf(strb, "%0.2f&mu;s",
```
Comment (brminich) [inline src/ucp/proto/proto_debug.c:643]: is this a specificator of `μs`?

File: src/ucp/proto/proto_debug.c [hunk 2]
```diff
@@ -609,6 +610,265 @@ const char *ucp_proto_perf_node_desc(ucp_proto_perf_node_t *perf_node)
...
+    khiter = kh_put(ucp_proto_graph_node, nodes_hash, perf_node, &khret);
+    if (khret == UCS_KH_PUT_KEY_PRESENT) {
+        id = kh_value(nodes_hash, khiter);
+        ucp_proto_perf_graph_str_append_link(parent_id, id, highlight, strb);
+        return;
+    } else if (khret == UCS_KH_PUT_FAILED) {
+        return;
+    }
+
+    id = kh_size(nodes_hash); /* We never remove form the hash */
```
Comment (brminich) [inline src/ucp/proto/proto_debug.c:696]: `from`

File: src/ucp/proto/proto_debug.c [hunk 3]
```diff
@@ -609,6 +610,265 @@ const char *ucp_proto_perf_node_desc(ucp_proto_perf_node_t *perf_node)
...
+        return;
+    }
+
+    khiter = kh_put(ucp_proto_graph_node, nodes_hash, perf_node, &khret);
+    if (khret == UCS_KH_PUT_KEY_PRESENT) {
+        id = kh_value(nodes_hash, khiter);
+        ucp_proto_perf_graph_str_append_link(parent_id, id, highlight, strb);
+        return;
+    } else if (khret == UCS_KH_PUT_FAILED) {
+        return;
```
Comment (brminich) [inline src/ucp/proto/proto_debug.c:693]: maybe `warning` or `diag`?

File: src/ucp/proto/proto_select.c [hunk 4]
```diff
@@ -317,6 +312,31 @@ ucp_proto_select_init_protocols(ucp_worker_h worker,
     return status;
 }
 
+static void
+ucp_proto_select_perf_ranges_cleanup(ucp_proto_perf_range_t *perf_ranges,
+                                     unsigned num_ranges)
+
```
Comment (brminich) [inline src/ucp/proto/proto_select.c:318]: plz remove

File: src/ucp/proto/proto_debug.c (hunk 1 above)
Comment (yosefe) [reply src/ucp/proto/proto_debug.c:643]: yes

**Review by brminich — APPROVED** [review]

**Pull request merged**